sound/soc/sof/topology.c

```diff
@@ -1219,24 +1267,48 @@ static int sof_widget_load_siggen(struct snd_soc_component *scomp, int index,
 			le32_to_cpu(private->size));
 		return ret;
 	}
+	dev_dbg(sdev->dev, "tone %s: frequency %d amplitude %d sample rate %d\n",
+		swidget->widget->name, tone.frequency, tone.amplitude,
+		tone.sample_rate);
 
-	ret = sof_parse_tokens(scomp, &tone.config, comp_tokens,
-			       ARRAY_SIZE(comp_tokens), private->array,
+	ret = sof_ipc_tx_message(sdev->ipc,
+				 tone.comp.hdr.cmd, &tone, sizeof(tone), r,
+				 sizeof(*r));
+	if (ret < 0)
+		return ret;
+
+	/* parse virtual FE link tokens */
+	ret = sof_parse_tokens(scomp, &vfe, vfe_tokens,
+			       ARRAY_SIZE(vfe_tokens), private->array,
 			       le32_to_cpu(private->size));
 	if (ret != 0) {
-		dev_err(sdev->dev, "error: parse tone.cfg tokens failed %d\n",
+		dev_err(sdev->dev, "error: parse vfe tokens failed %d\n",
 			le32_to_cpu(private->size));
 		return ret;
 	}
 
-	dev_dbg(sdev->dev, "tone %s: frequency %d amplitude %d sample rate %d\n",
-		swidget->widget->name, tone.frequency, tone.amplitude,
-		tone.sample_rate);
-	sof_dbg_comp_config(scomp, &tone.config);
+	dev_dbg(sdev->dev,
+		"VFE link name %s: cpu dai name %s platform name %s\n",
+		vfe.link_name, vfe.cpu_dai_name, vfe.platform_name);
 
-	return sof_ipc_tx_message(sdev->ipc,
-				  tone.comp.hdr.cmd, &tone, sizeof(tone), r,
-				  sizeof(*r));
+	/*
+	 * create virtual FE link
+	 * This will be used to enable the codec when
+	 * the siggen pipeline is triggered
+	 */
+	ret = snd_soc_dpcm_vfe_new(scomp->card, index, vfe.link_name,
+				   vfe.cpu_dai_name, vfe.platform_name,
+				   SNDRV_PCM_STREAM_PLAYBACK);
```

Comment:
I think we should make these vfe either playback or capture - with an addition to dapm to mirror the siggen functionality for capture

Reply:
The VFE is already defined such that it is playback or capture. The last argument in the vfe_new() function defines the stream direction for the VFE. In the case of siggen, I specify the direction for playback .

I will add a new dapm widget for siggen_sink that will be used for capture.